storage/indexer/src/event_v2_translator.rs

```diff
@@ -0,0 +1,1327 @@
+        let object_group_resource: ObjectGroupResource = bcs::from_bytes(state_value.bytes())?;
+        Ok(object_group_resource
+            .group
+            .get(struct_tag)
+            .map(|bytes| Bytes::copy_from_slice(bytes)))
+    }
+}
+
+struct CoinDepositTranslator;
+impl EventV2Translator for CoinDepositTranslator {
+    fn translate_event_v2_to_v1(
+        &self,
+        v2: &ContractEventV2,
+        engine: &EventV2TranslationEngine,
+    ) -> Result<ContractEventV1> {
+        let coin_deposit = CoinDeposit::try_from_bytes(v2.event_data())?;
+        let struct_tag_str = format!("0x1::coin::CoinStore<{}>", coin_deposit.coin_type());
+        let struct_tag = StructTag::from_str(&struct_tag_str)?;
+        let (key, sequence_number) = if let Some(state_value_bytes) =
+            engine.get_state_value_bytes_for_resource(coin_deposit.account(), &struct_tag)?
+        {
+            // We can use `DummyCoinType` as it does not affect the correctness of deserialization.
+            let coin_store_resource: CoinStoreResource<DummyCoinType> =
+                bcs::from_bytes(&state_value_bytes)?;
+            let key = *coin_store_resource.deposit_events().key();
+            let sequence_number = engine
+                .get_next_sequence_number(&key, coin_store_resource.deposit_events().count())?;
+            (key, sequence_number)
+        } else {
+            (EventKey::new(2, *coin_deposit.account()), 0)
+        };
+        let deposit_event = DepositEvent::new(coin_deposit.amount());
+        Ok(ContractEventV1::new(
+            key,
+            sequence_number,
+            DEPOSIT_EVENT_TYPE.clone(),
+            bcs::to_bytes(&deposit_event)?,
+        ))
+    }
+}
+
+struct CoinWithdrawTranslator;
+impl EventV2Translator for CoinWithdrawTranslator {
+    fn translate_event_v2_to_v1(
+        &self,
+        v2: &ContractEventV2,
+        engine: &EventV2TranslationEngine,
+    ) -> Result<ContractEventV1> {
+        let coin_withdraw = CoinWithdraw::try_from_bytes(v2.event_data())?;
+        let struct_tag_str = format!("0x1::coin::CoinStore<{}>", coin_withdraw.coin_type());
+        let struct_tag = StructTag::from_str(&struct_tag_str)?;
+        let (key, sequence_number) = if let Some(state_value_bytes) =
+            engine.get_state_value_bytes_for_resource(coin_withdraw.account(), &struct_tag)?
+        {
+            // We can use `DummyCoinType` as it does not affect the correctness of deserialization.
+            let coin_store_resource: CoinStoreResource<DummyCoinType> =
+                bcs::from_bytes(&state_value_bytes)?;
+            let key = *coin_store_resource.withdraw_events().key();
+            let sequence_number = engine
+                .get_next_sequence_number(&key, coin_store_resource.withdraw_events().count())?;
+            (key, sequence_number)
+        } else {
+            (EventKey::new(3, *coin_withdraw.account()), 0)
```

Comment:
what is 3 here?

Reply:
It's the creation number of the event key for `WithdrawEvent`. It's deterministically 3.

Reply:
Can you define some constant here for better readability?

Reply:
I defined the corresponding constants, adding some comments.